### ISSUE TITLE
ServerSide Diff failing for CRs when a new field is added to the CRD: field not declared in schema

### ISSUE THREAD

**Issue by varmarohit75** [argoproj/argo-cd#20458]
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq. 
SImilar to https://github.com/argoproj/argo-cd/issues/15176 but for CRD caching
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

We are using ArgoCD v2.10.9 to manage several CRDs deployed in our Kubernetes cluster. 
Additionally, we have other applications that deploy CRs based on these CRDs through ArgoCD as well.

Whenever a new field is added to an existing CRD, it is successfully applied to the cluster, and all existing CRs are reconciled without issues. However, we encounter the following problem when trying to apply a CR with the updated CRD through ArgoCD:

- Error on ArgoCD Sync:
During the serverside diff operation, we receive the following error:
```
ComparisonError: Failed to compare desired state to live state: failed to calculate diff from cache: error calculating server side diff:
serverSideDiff error: error removing non config mutations for resource CustomStatefulSet/test-app-1:
error converting predicted live state from unstructured to example.com/v1alpha1, Kind=CustomStatefulSet: .spec.application.resourceRequirements: field not declared in schema

```

- Similar error is seen on Hard Refresh and sync:

```
ComparisonError:  Failed to compare desired state to live state: failed to calculate diff: error calculating server side diff:
serverSideDiff error: error removing non config mutations for resource CustomStatefulSet/test-app-1:
error converting predicted live state from unstructured to example.com/v1alpha1, Kind=CustomStatefulSet: .spec.application.resourceRequirements: field not declared in schema

```

**To Reproduce**

- Apply a CRD through ArgoCD with ServerSide diff set to true.
- Apply corresponding CRs for that CRD using ArgoCD.
- Add a new field to the CRD (preferably with a default value).
- Ensure that the new field is updated across all CRs in the Kubernetes cluster.
- Sync the CR in ArgoCD—this is where the error should occur.

**Expected behavior**

The cache to be reset timely when the updated CRD is applied.
Or when a sync is performed on the CR

**Version**

```
argocd: v2.11.3+3f344d5
  BuildDate: 2024-06-06T12:31:55Z
  GitCommit: 3f344d54a4e0bbbb4313e1c19cfe1e544b162598
  GitTreeState: clean
  GoVersion: go1.22.4
  Compiler: gc
  Platform: darwin/arm64
argocd-server: v2.10.9+c071af8
  BuildDate: 2024-04-30T15:53:28Z
  GitCommit: c071af808170bfc39cbdf6b9be4d0212dd66db0c
  GitTreeState: clean
  GoVersion: go1.21.3
  Compiler: gc
  Platform: linux/amd64
  Kustomize Version: v5.2.1 2023-10-19T20:13:51Z
  Helm Version: v3.14.3+gf03cc04
  Kubectl Version: v0.26.11
  Jsonnet Version: v0.20.0
 
 ```

**Logs**

```
msg="Applying resource CustomStatefulSet/erbium-seas-cloud-searcher-i002 in cluster: https://10.244.0.1:443, namespace: seas-cloud-searcher-6a8f2" dry-run=server manager=argocd-controller serverSideApply=true serverSideDiff=true

msg="Normalized app spec: {\"status\":{\"conditions\":[{\"lastTransitionTime\":\"2024-10-17T19:40:50Z\",\"message\":\"Failed to compare desired state to live state: failed to calculate diff from cache: error calculating server side diff: serverSideDiff error: error removing non config mutations for resource CustomStatefulSet/erbium-seas-cloud-searcher-i002: error converting predicted live state from unstructured to example.com/v1alpha1, Kind=CustomStatefulSet: .spec.application.resourceRequirements: field not declared in schema\",\"type\":\"ComparisonError\"}]}}" application=argocd/seas-cloud-searcher
```

**Mitigated through**

A rollout restart of statefulset seems to fix the issue and the sync on CR goes through

```
kubectl rollout restart statefulset argocd-application-controller -n argocd
```

**Comment by andrii-korotkov-verkada** [comment]
What's this `.spec.application.resourceRequirements` field? Looks like Kubernetes might complain as well if you try to apply directly. This sounds like an error in the application configuration. Try to remove the field or move to where it should belong.

**Comment by andrii-korotkov-verkada** [comment]
If this doesn't solve the issue, try upgrading to v2.12 or v2.13, there's been a lot of improvements done.

**Comment by varmarohit75** [comment]
> What's this .spec.application.resourceRequirements field? 

This field was recently added to our CRD as part of an update that was applied successfully to the cluster. As a result, custom resources in the cluster now reflect this new field. However, when we trigger a sync from ArgoCD, the server-side diff returns an error, indicating that it doesn’t recognize the field.

**Comment by andrii-korotkov-verkada** [comment]
Would you be able to build an image from master and check if the error persists with server side diff on?
There's been a re-work of how server side diff works https://github.com/argoproj/gitops-engine/commit/4fd18478f5946b756fe253055358f89d4dd403a9, which should have made things better.

**Comment by zoetrope** [comment]
I have encountered the same issue.

- I am able to reproduce this problem using the latest Argo CD (v2.14.10).
- The issue occurs even when ServerSideApply is not enabled. In that case, no error message is shown, but the resource remains in the OutOfSync state.
- I suspect that the following schema update process is not being triggered until the Sync operation is completed: https://github.com/argoproj/gitops-engine/blob/5f90e7b481b787e43cb53148a5d78a8dd04e5ce5/pkg/cache/cluster.go#L895

**Comment by chiragjn** [comment]
Bump on this issue, we face this issue often with Nvidia's gpu-operator chart where new CR in the chart is compared with the old CRD.
The chart has proper helm hooks (job) to upgrade the CRD but argocd won't start Pre-sync phase to let the job run :/ 

We are on argocd 2.14.10

**Comment by ahmetb** [comment]
I've managed to get a similar error (`unable to resolve parseableType for GroupVersionKind`) on a builtin k8s type (not CR) after a k8s upgrade on argo-cd v2.14.9 (38985bd).

```
Failed to compare desired state to live state: failed to calculate diff:
error calculating server side diff: serverSideDiff error:
error removing non config mutations for resource PriorityLevelConfiguration/my-name:
unable to resolve parseableType for GroupVersionKind: flowcontrol.apiserver.k8s.io/v1beta3,
Kind=PriorityLevelConfiguration
```

Repro steps:
- I had a PriorityLevelConfiguration v1beta2 resource in git, deployed on k8s v1.25.
- I upgraded apiserver to k8s 1.28 (which offers both v1beta2+v1beta3 for this resource)
- I changed PriorityLevelConfiguration resource to v1beta3 in git + triggered a sync
  - Note: v1beta2 → v1beta3 migration for the PriorityLevelConfiguration API involves a field name change
- Got the pasted error during sync.
- Initially did `kubectl rollout restart` but for whatever reason that didn't work until like an hour later and multiple restarts, so not sure if another component was caching the GVK info.
- Eventually worked fine after some restarts.

**Comment by pjiang-dev** [comment]
@ahmetb @varmarohit75
Could you provide the CRD schema? 

The server side diff will have trouble if there are schema validation issues. Particularly if CRD has CRD has `preserveUnknownFields: true`

Something like this command will tell if there are issues with schema
`kubectl get crd iamroles.iammanager.keikoproj.io -o yaml`

**Comment by ahmetb** [comment]
@pjiang-dev as I said it's a builtin type; not a CRD.

**Comment by pjiang-dev** [comment]
@ahmetb Yes I think this may be a cache issue, as we don't always get the latest live resource when doing server side diff. I will investigate it further

**Comment by pjiang-dev** [comment]
Tested exactly the steps in the original issue and i am not able to reproduce on `v3.2.0+`
